### ISSUE TITLE
Under WSL, torch.xpu.clear_cache() will leak VRAM in specific circumstances

### ISSUE THREAD

**Issue by a-One-Fan** [intel/intel-extension-for-pytorch#388]
### Describe the bug

Doing torch.xpu.clear_cache() will leak VRAM in WSL under specific circumstances. It's not clear to me what actually triggers the leaking beyond "there's complex stuff going on" or maybe that there's some cache that actually needs to be cleared. Either way, the issue happens with trying to clear cache in the Automatic1111 Stable Diffusion, or with Tortoise. This does not happen on native Linux, only on WSL. The leaking can easily get very severe, the more calls to clear_cache the more severe it will get. I have trimmed down a single ~600 line file that reproduces the issue.

I have tested with the 4514 driver on an A770 16GB.

The file will make a folder in your current working directory, and download a 1.7GB model to it. You will need `progressbar` and `transformers==4.19`.

[memleak.txt](https://github.com/intel/intel-extension-for-pytorch/files/12017022/memleak.txt)

Near the bottom of the file, in the inference loop is the torch.xpu.clear_cache. Commenting it out, no leak will happen. With it there, IPEX will report almost zero increase in VRAM, but Task Manager will show it steadily climbing (what seems like ~20mb/s), until for this specific case, at ~7GB (again: I have 16GB) IPEX will OOM. In other cases, it can easily leak to bigger 15-16GB and have a less or more serious OOM then.

```
  File "/home/vik/vc/./memleak.py", line 587, in <module>
  File "/home/vik/vc/./memleak.py", line 569, in tts
  File "/home/vik/vc/./memleak.py", line 463, in inference_speech
  File "/home/vik/miniconda3/envs/test2/lib/python3.10/site-packages/torch/autograd/grad_mode.py", line 27, in decorate_context
  File "/home/vik/miniconda3/envs/test2/lib/python3.10/site-packages/transformers/generation_utils.py", line 1310, in generate
  File "/home/vik/miniconda3/envs/test2/lib/python3.10/site-packages/transformers/generation_utils.py", line 1926, in sample
  File "/home/vik/miniconda3/envs/test2/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1194, in _call_impl
  File "/home/vik/vc/./memleak.py", line 253, in forward
  File "/home/vik/miniconda3/envs/test2/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1194, in _call_impl
  File "/home/vik/miniconda3/envs/test2/lib/python3.10/site-packages/transformers/models/gpt2/modeling_gpt2.py", line 889, in forward
  File "/home/vik/miniconda3/envs/test2/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1194, in _call_impl
  File "/home/vik/miniconda3/envs/test2/lib/python3.10/site-packages/transformers/models/gpt2/modeling_gpt2.py", line 390, in forward
  File "/home/vik/miniconda3/envs/test2/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1194, in _call_impl
  File "/home/vik/miniconda3/envs/test2/lib/python3.10/site-packages/transformers/models/gpt2/modeling_gpt2.py", line 321, in forward
RuntimeError: DPCPP out of memory. Tried to allocate 2.00 MiB (GPU
```
It is cut off like that.

### Versions

Collecting environment information...
PyTorch version: 1.13.0a0+git6c9b55e
PyTorch CXX11 ABI: Yes
IPEX version: 1.13.120+xpu
IPEX commit: c2a37012e
Build type: Release

OS: Ubuntu 22.04.2 LTS (x86_64)
GCC version: (Ubuntu 11.3.0-1ubuntu1~22.04.1) 11.3.0
Clang version: N/A
IGC version: 2023.1.0 (2023.1.0.20230320)
CMake version: N/A
Libc version: glibc-2.35

Python version: 3.10.12 (main, Jul  5 2023, 18:54:27) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.15.90.1-microsoft-standard-WSL2-x86_64-with-glibc2.35
Is XPU available: True
DPCPP runtime version: 2023.1.0
MKL version: 2023.1.0
GPU models and configuration:
[0] _DeviceProperties(name='Intel(R) Graphics [0x56a0]', platform_name='Intel(R) Level-Zero', dev_type='gpu, support_fp64=0, total_memory=13004MB, max_compute_units=512)
Intel OpenCL ICD version: 23.17.26241.21-647~22.04
Level Zero version: 1.3.26241.21-647~22.04

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Address sizes:                   48 bits physical, 48 bits virtual
Byte Order:                      Little Endian
CPU(s):                          16
On-line CPU(s) list:             0-15
Vendor ID:                       AuthenticAMD
Model name:                      AMD Ryzen 7 3700X 8-Core Processor
CPU family:                      23
Model:                           113
Thread(s) per core:              2
Core(s) per socket:              8
Socket(s):                       1
Stepping:                        0
BogoMIPS:                        7186.53
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl tsc_reliable nonstop_tsc cpuid extd_apicid pni pclmulqdq ssse3 fma cx16 sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm cmp_legacy svm cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw topoext perfctr_core ssbd ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves clzero xsaveerptr arat npt nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold v_vmsave_vmload umip rdpid
Virtualization:                  AMD-V
Hypervisor vendor:               Microsoft
Virtualization type:             full
L1d cache:                       256 KiB (8 instances)
L1i cache:                       256 KiB (8 instances)
L2 cache:                        4 MiB (8 instances)
L3 cache:                        16 MiB (1 instance)
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Not affected
Vulnerability Retbleed:          Mitigation; untrained return thunk; SMT enabled with STIBP protection
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Retpolines, IBPB conditional, STIBP always-on, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected

Versions of relevant libraries:
[pip3] intel-extension-for-pytorch==1.13.120+xpu
[pip3] numpy==1.25.1
[pip3] torch==1.13.0a0+git6c9b55e
[conda] intel-extension-for-pytorch 1.13.120+xpu             pypi_0    pypi
[conda] numpy                     1.25.1                   pypi_0    pypi
[conda] torch                     1.13.0a0+git6c9b55e          pypi_0    pypi

**Comment by a-One-Fan** [comment]
> Same [here](https://github.com/Nuullll/ipex-sd-docker-for-arc-gpu/issues/8). This is a severe problem that makes IPEX almost unusable in WSL.

Yeah, this is that VRAM issue. For months now people knew there's this issue, and I assumed it was common knowledge due to how prevalent and severe it is, but I guess not.
clear_cache is most likely the main underlying problem. I asked someone else, and commenting out the clear_cache in their copy of jbaboval's Autoamtic1111 SD fork fixed their leaks. I'm not certain if this is ALL the leaks there are for now, however this should be a decent chunk of them. 
As another alternative, you can also use diffusers, which will not leak (I'm pretty sure it does not clear_cache).

**Comment by jingxu10** [comment]
@gujinghui

**Comment by Nuullll** [comment]
https://github.com/Nuullll/ipex-sd-docker-for-arc-gpu/issues/8#issuecomment-1632155553 : confirmed that `torch.xpu.empty_cache()` was causing huge amount of VRAM leak.

**Comment by gujinghui** [comment]
> This does not happen on native Linux, only on WSL.

Sounds like it's not issue in IPEX.
It should be issue in runtime on WSL, and driver on windows...

@sanchitintel  @min-jean-cho Can you help to check this issue?
